modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,337 @@
+  params.addRequiredParam<subdomain_id_type>(
+      "peripheral_ring_block_id", "The block id assigned to the created peripheral layer.");
+  params.addParam<SubdomainName>("peripheral_ring_block_name",
+                                 "The block name assigned to the created peripheral layer.");
+  params.addRangeCheckedParam<boundary_id_type>("external_boundary_id",
+                                                "external_boundary_id>0",
+                                                "Optional customized external boundary id.");
+  params.addParam<std::string>("external_boundary_name",
+                               "Optional customized external boundary name.");
+  params.addClassDescription("This PeripheralRingMeshGenerator object adds a circular peripheral "
+                             "region to the input mesh.");
+
+  return params;
+}
+
+PeripheralRingMeshGenerator::PeripheralRingMeshGenerator(const InputParameters & parameters)
+  : PolygonMeshGeneratorBase(parameters),
+    _input_name(getParam<MeshGeneratorName>("input")),
+    _peripheral_layer_num(getParam<unsigned int>("peripheral_layer_num")),
+    _peripheral_ring_radius(getParam<Real>("peripheral_ring_radius")),
+    _preserve_volumes(getParam<bool>("preserve_volumes")),
+    _input_mesh_external_boundary(getParam<BoundaryName>("input_mesh_external_boundary")),
+    _peripheral_ring_block_id(getParam<subdomain_id_type>("peripheral_ring_block_id")),
+    _peripheral_ring_block_name(isParamValid("peripheral_ring_block_name")
+                                    ? getParam<SubdomainName>("peripheral_ring_block_name")
+                                    : (SubdomainName) ""),
+    _external_boundary_id(isParamValid("external_boundary_id")
+                              ? getParam<boundary_id_type>("external_boundary_id")
+                              : 0),
+    _external_boundary_name(isParamValid("external_boundary_name")
+                                ? getParam<std::string>("external_boundary_name")
+                                : std::string()),
+    _mesh(getMeshByName(_input_name))
+{
+}
+
+std::unique_ptr<MeshBase>
+PeripheralRingMeshGenerator::generate()
+{
+  // Need ReplicatedMesh for stitching
+  auto input_mesh = dynamic_cast<ReplicatedMesh *>(_mesh.get());
+  _input_mesh_external_bid =
+      MooseMeshUtils::getBoundaryID(_input_mesh_external_boundary, *input_mesh);
+
+  // Use CoM of the input mesh as its origin for azimuthal calculation
+  const Point origin_pt = MooseMeshUtils::meshCentroidCalculator(*input_mesh);
+  const Real origin_x = origin_pt(0);
+  const Real origin_y = origin_pt(1);
+  // Vessel for containing maximum radius of the boundary nodes
+  Real max_input_mesh_node_radius;
+
+  if (!isBoundaryValid(
+          *input_mesh, max_input_mesh_node_radius, origin_pt, _input_mesh_external_bid))
+    paramError("input_mesh_external_boundary",
+               "This mesh generator does not work for the provided boundary.");

```

Comment:
let's try to add a list of possible reasons there so a user has somewhere to go from there